Fix Xiaomi BLE not detecting encryption for some devices

Jc2k opened the pull request:
## Proposed change

When a Xiaomi MiBeacon device sends its advertisements it can send packets that don't have a payload. As there is no payload, there is no encryption, and if there is no encryption we can't get the config flow to prompt for a bindkey.

After discussing with @bdraco, the best plan right now is to wait in the config flow until we see an advertisement matching our requirements OR we time out.

I broke out the bluetooth framework part of this code into a helper in case it would be useful to other integrations, but it can live in xiaomi_ble if thats preferred.

## Type of change
<!--
  What type of change does your PR introduce to Home Assistant?
  NOTE: Please, check only 1! box!
  If your PR requires multiple boxes to be checked, you'll most likely need to
  split it into multiple PRs. This makes things easier and faster to code review.
-->

- [ ] Dependency upgrade
- [x] Bugfix (non-breaking change which fixes an issue)
- [ ] New integration (thank you!)
- [ ] New feature (which adds functionality to an existing integration)
- [ ] Deprecation (breaking change to happen in the future)
- [ ] Breaking change (fix/feature causing existing functionality to break)
- [ ] Code quality improvements to existing code or addition of tests

## Additional information
<!--
  Details are important, and help maintainers processing your PR.
  Please be sure to fill out additional details, if applicable.
-->

- This PR fixes or closes issue: fixes #75833
- This PR is related to issue: 
- Link to documentation pull request: 

## Checklist
<!--
  Put an `x` in the boxes that apply. You can also fill these out after
  creating the PR. If you're unsure about any of them, don't hesitate to ask.
  We're here to help! This is simply a reminder of what we are going to look
  for before merging your code.
-->

- [ ] The code change is tested and works locally.
- [ ] Local tests pass. **Your PR cannot be merged unless tests pass**
- [ ] There is no commented out code in this PR.
- [ ] I have followed the [development checklist][dev-checklist]
- [ ] The code has been formatted using Black (`black --fast homeassistant tests`)
- [ ] Tests have been added to verify that the new code works.

If user exposed functionality or configuration variables are added/changed:

- [ ] Documentation added/updated for [www.home-assistant.io][docs-repository]

If the code communicates with devices, web services, or third-party tools:

- [ ] The [manifest file][manifest-docs] has all fields filled out correctly.  
      Updated and included derived files by running: `python3 -m script.hassfest`.
- [ ] New or updated dependencies have been added to `requirements_all.txt`.  
      Updated by running `python3 -m script.gen_requirements_all`.
- [ ] For the updated dependencies - a link to the changelog, or at minimum a diff between library versions is added to the PR description.
- [ ] Untested files have been added to `.coveragerc`.

The integration reached or maintains the following [Integration Quality Scale][quality-scale]:
<!--
  The Integration Quality Scale scores an integration on the code quality
  and user experience. Each level of the quality scale consists of a list
  of requirements. We highly recommend getting your integration scored!
-->

- [ ] No score or internal
- [ ] 🥈 Silver
- [ ] 🥇 Gold
- [ ] 🏆 Platinum

<!--
  This project is very active and we have a high turnover of pull requests.

  Unfortunately, the number of incoming pull requests is higher than what our
  reviewers can review and merge so there is a long backlog of pull requests
  waiting for review. You can help here!
  
  By reviewing another pull request, you will help raise the code quality of
  that pull request and the final review will be faster. This way the general
  pace of pull request reviews will go up and your wait time will go down.
  
  When picking a pull request to review, try to choose one that hasn't yet
  been reviewed.

  Thanks for helping out!
-->

To help with the load of incoming pull requests:

- [ ] I have reviewed two other [open pull requests][prs] in this repository.

[prs]: https://github.com/home-assistant/core/pulls?q=is%3Aopen+is%3Apr+-author%3A%40me+-draft%3Atrue+-label%3Awaiting-for-upstream+sort%3Acreated-desc+review%3Anone+-status%3Afailure

<!--
  Thank you for contributing <3

  Below, some useful links you could explore:
-->
[dev-checklist]: https://developers.home-assistant.io/docs/en/development_checklist.html
[manifest-docs]: https://developers.home-assistant.io/docs/en/creating_integration_manifest.html
[quality-scale]: https://developers.home-assistant.io/docs/en/next/integration_quality_scale_index.html
[docs-repository]: https://github.com/home-assistant/home-assistant.io

Jc2k committed:
Add 'async_process_advertisements' and update config flow to use it

Jc2k committed:
Unused import

MartinHjelmare commented on homeassistant/components/xiaomi_ble/config_flow.py:
```diff
@@ -53,6 +72,12 @@ async def async_step_bluetooth(
             return self.async_abort(reason="not_supported")
+
+        # Wait until we have received enough information about this device to detect its encryption type
+        await self._async_wait_for_full_advertisement(discovery_info, device)
```
We need to catch `asyncio.TimeoutError` here, right?

Jc2k replied:
Probably! I'm just working on tests now to make sure i'm happy with the API of the helpers. I think I will let them raise, and then abort the config flow on the TimeoutError, rather than checking device.pending again (If there is a Timeout it will still be pending, if there is not then it bailed early and hence is not pending).

Jc2k replied:
Latest push does this and has tests for both step_user and step_bluetooth where the timeout expires and also where a suitable advertisement is injected, allowing the flow to recover and correctly detect the encryption type.

Jc2k committed:
Use a const so we can patch it in tests

Jc2k committed:
Handle timeout error and abort config flow

Jc2k committed:
Test that config flow is aborted early if no additional packets received

Jc2k committed:
Test can recover when later advertisements have neccesary data

Jc2k committed:
Finish tests

Jc2k committed:
Fix tests on linux

On Linux we are able to compare the MAC from bluez to the MAC that
is (sometimes) baked into the service data. For this test I had to
create fake test data, and didn't update the MAC that was embedded
in the service_data.

macOS doesn't do this check as we don't have the raw MAC field to
compare to.

bdraco reviewed (approved):
Passing on linux now 👍
